tests/modules/keyspecs.c

```diff
@@ -23,92 +23,139 @@ int kspec_complex2(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
     return REDISMODULE_OK;
 }
 
-int RedisModule_OnLoad(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
-    REDISMODULE_NOT_USED(argv);
-    REDISMODULE_NOT_USED(argc);
-
-    int spec_id;
-
-    if (RedisModule_Init(ctx, "keyspecs", 1, REDISMODULE_APIVER_1)== REDISMODULE_ERR)
-        return REDISMODULE_ERR;
-
+int createKspecLegacy(RedisModuleCtx *ctx) {
     /* Test legacy range "gluing" */
     if (RedisModule_CreateCommand(ctx,"kspec.legacy",kspec_legacy,"",0,0,0) == REDISMODULE_ERR)
         return REDISMODULE_ERR;
```

Comment:
i don't understand how we test here the legacy range gluing if the command is registered with 0,0,0 (no keys)?

correct me if i'm wrong, but the idea is that simple 1 key commands don't need to declare key-specs at all (unless they wanna provide flags).

i suppose we need to test these:
1. module that's only using the basic RM_CreateCommand. we wanna test that COMMAND INFO does show one spec (without flags).
2. module that's using RM_CreateCommand to register the spec (to be backwards compatible with old redis versions), and later provides the same spec using RM_SetCommandInfo to add the flags.
3. module that declares one spec using RM_CreateCommand and adds another using RM_CreateCommand (i.e. gluing, this is what `createKspecComplex1` below seems to test)

i suppose we also wanna document this (provide recommendation on how it should be used).
@guybe7 @yossigo please provide feedback.

p.s. when combining this with the `getkeys-api` flag and the `RM_KeyAtPos`, this topic is a complete mess.

Reply:
@guybe7 you wrote these tests originally? Maybe you know what the idea was? I merely converted them to the declarative style.

Reply:
i think the idea was to make sure specs are merged (i remember discussing case 3 with him), but looking at the test i think it's missing part of the point.
please extend the tests to cover all these cases.

Reply:
@oranagra Is there a real use case for [3]?
Regarding `getkeys-api`, maybe it's time to deprecate it? Possibly produce a warning if command info is attached to a command that has it enabled? I realize some commands may not be keyspecs-compatible but in the long term we do want module developers to align with keyspecs.

Reply:
let me try to analyze two separate concerns:

1. what modules had to do for redis 6.0.
  For 6.0, modules had to use both the initial triplet (for cluster aware clients), and also the getkeys-api (for everything else, inside redis, and COMMAND GETKEYS)
2. what we expect modules to do for 7.0.
  for 7.0, we can maybe rely only on key-specs, but the fact is that they're a lot more work to declare, so maybe for simple modules, that require only one key, we can let them keep using the initial triplet (in theory, that can serve them for anything, including ACL, except for cases where ACL is used with explicit read or write permissions).

Note however that this is not yet the case. current code in unstable still replies on getkeys-api for anything other than ACL (and COMMAND command).
i.e. it uses that for COMMAND GETKEYS, client side tracking, cluster decisions, and RM_GetCommandKeys.

Modules that want to support two versions of redis, will be able to implement both methods, i.e. use the old API, and also the new API when it's available.
so the use case for [3] is that a module which was written for redis 6, has to declare the initial triplet (for cluster aware clients), but also use the `getkeys-api` in for the sake of well, anything else. and declare key-specs for 7.0 ACL.
I don't see any way around this since ACL (at least per key read and write permissions) must rely on key-specs, and in order to support redis 6.0, the module must provide both getkeys-api and the initial triplet.

Reply:
@oranagra This makes sense. But what about option [3] above? Why would a module provide an incomplete triplet and later pass that using command info?

Reply:
@yossigo it'll pass the first triplet to be compatible with redis 6.0 and cluster aware clients.
it can later pass the same triplet again in key-specs in order to provide flags.
and it can pass additional specs (e.g. a STORE argument) which it couldn't expose before.

I don't recall what the "gluing" term was about (in the original key-specs PR).
it could have meant to glue two adjacent range specs into one.

Reply:
btw, i spoke with Guy about that test and the "gluing" part.
the test attempted to test that two separate range key-specs that are defined with the new API (e.g. each one with different flags), are shown in the legacy part of COMMAND command output as one range.

i.e. the gluing is not about gluing key-specs together, or gluing them with what's defined in RM_CreateCommand.
it's about the output of the legacy part of COMMAND command.

Reply:
OK. Conclusion? Should I add some comments?

Reply:
i think
1. add comments about the gluing and what this test verifies (the legacy response of COMMAND), please verify that the test really does that.
2. add comments in the gluing code that specifies that too (that it's for the output of COMMAND command)
3. we don't need the registration of key-specs to amend what's given in the first triplet. but we do need it to be able to  override it (to add flags), please add a test for it.
4. we do need to document that both key-specs and getkeys-api are needed.
5. we do need to document that the triplet for RM_CreateCommand is needed if you wanna also support old versions of redis.

Reply:
> i think
> 
> 1. add comments about the gluing and what this test verifies (the legacy response of COMMAND), please verify that the test really does that.

Done.

> 2. add comments in the gluing code that specifies that too (that it's for the output of COMMAND command)

There is already a very detailed comment at the definition of `populateCommandLegacyRangeSpec` in server.c. It's used by non-module commands too. It starts like this: *The purpose of this function is to try to "glue" consecutive range key specs in order to build the legacy (first,last,step) spec used by the COMMAND command. (...)*

I added a brief comment at the call to this function in module.c too, to explain what it does.

> 3. we don't need the registration of key-specs to amend what's given in the first triplet. but we do need it to be able to  override it (to add flags), please add a test for it.

If you by "amend" mean combine information from the triplet with keyspecs, then I'm with you. The legacy triple is simply overridden by `populateCommandLegacyRangeSpec`, which is called if any key specs are provided. 

This test is verifying that, isn't it? If not, I don't see what test you're asking for here. Updating the legacy triple is done in the same way for module commands as for builtin commands.

> 4. we do need to document that both key-specs and getkeys-api are needed.

Adding this in the RM_SetCommandInfo docs: "Note that key-specs don't fully replace the "getkeys-api" (see RM_CreateCommand, RM_IsKeysPositionRequest and RM_KeyAtPos) so it may be a good idea to supply both key-specs and a implement the getkeys-api."

> 5. we do need to document that the triplet for RM_CreateCommand is needed if you wanna also support old versions of redis.

Adding this in the RM_SetCommandInfo docs: "Key-specs cause the triplet (firstkey, lastkey, keystep) given in RM_CreateCommand to be recomputed, but it is still useful to provide these three parameters in RM_CreateCommand, to better support old Redis versions where RM_SetCommandInfo is not available."

Reply:
> If you by "amend" mean combine information from the triplet with keyspecs, then I'm with you.

by "amend" i meant glue specs that seem consecutive into one.

> The legacy triple is simply overridden by populateCommandLegacyRangeSpec, which is called if any key specs are provided.

too lazy / busy to check.. is it overridden just in case they're matching (range spec with the same first/last/step)?
or does any spec added by key-specs overrides what's defined in command creation?
i prefer to override only if they're matching.

> This test is verifying that, isn't it? If not, I don't see what test you're asking for here. Updating the legacy triple is done in the same way for module commands as for builtin commands.

haven't looked at the test yet.
i think we should verify:
1. a key-spec that's added which is the same as the one used in command creation (being overwritten).
2. a key-spec that's added in addition to the one in command creation.
3. in both cases we need to check the key-specs in COMMAND output (and their flags), but also the legacy output of COMMAND command.

Reply:
I'm reluctant to treating module commands differently than native commands. This is the function in server.c that populates the (first,last,step) triple for all native Redis commands. (It's called by populateCommandTable which infers everything that isn't specified in in the command JSON data).

TL;DR: It glues consecutive keyspec ranges into one triple when it's strait-forward to do so. If it isn't possible, the function leaves the legacy triple unchanged. Thus, it's always safe to call it, regardles of whether an explicit legacy triple has been set or not.

```C
/* The purpose of this function is to try to "glue" consecutive range
 * key specs in order to build the legacy (first,last,step) spec
 * used by the COMMAND command.
 * By far the most common case is just one range spec (e.g. SET)
 * but some commands' ranges were split into two or more ranges
 * in order to have different flags for different keys (e.g. SMOVE,
 * first key is "read write", second key is "write").
 *
 * This functions uses very basic heuristics and is "best effort":
 * 1. Only commands which have only "range" specs are considered.
 * 2. Only range specs with keystep of 1 are considered.
 * 3. The order of the range specs must be ascending (i.e.
 *    lastkey of spec[i] == firstkey-1 of spec[i+1]).
 *
 * This function will succeed on all native Redis commands and may
 * fail on module commands, even if it only has "range" specs that
 * could actually be "glued", in the following cases:
 * 1. The order of "range" specs is not ascending (e.g. the spec for
 *    the key at index 2 was added before the spec of the key at
 *    index 1).
 * 2. The "range" specs have keystep >1.
 *
 * If this functions fails it means that the legacy (first,last,step)
 * spec used by COMMAND will show 0,0,0. This is not a dire situation
 * because anyway the legacy (first,last,step) spec is to be deprecated
 * and one should use the new key specs scheme.
 */
void populateCommandLegacyRangeSpec(struct redisCommand *c) {
```

Only one thing above is incorrect: it doesn't set the triple to (0,0,0) on failure. It simply leaves the triple unchanged on failure. (It matters only if it had any value before this function was called).

Reply:
@zuiderkwast the difference between modules and native redis commands, is that native command don't declare this triple, they only declare key-specs, and this code populates that triple for the benefit of the output of COMMAND command.
An example is LMOVE which has two key-specs, that are glued together for a single range triple in COMMAND output.

for module commands, they do provide the triple, so only here we're facing the dilemma of overwriting or gluing.
since we do want modules (e.g. ones with a single key argument) to use both (one for compatibility with old redis versions, and the other for the key-spec flags), we need to handle an overwrite case.

i.e two examples and how i think they should be handled:
1. a command with a single key argument. will use the initial triple for old version compat, and a key-spec for new redis versions (with flags), in this case we certainly wanna override.
2. a module with a command like LMOVE, may be using a 2 key triple in the command registration, and then define two key-specs. i suppose the first spec can override the initial triple, and then the second spec be glued (for a 2 key output in the legacy triple of COMMAND command)

Reply:
@oranagra It's seems that you disagree (for the sake of arguing? :grin:) but in the last example we actually seem to agree.

> 2. a module with a command like LMOVE, may be using a 2 key triple in the command registration, and then define two key-specs. i suppose the first spec can override the initial triple, and then the second spec be glued (for a 2 key output in the legacy triple of COMMAND command)

This can also be seen as both specs together override the triple. Isn't this gluing (range-concatenation) and amending the same thing?

Reply:
not for the sake of arguing. trying to break it down and figure out what's right.

i the beginning of this discussion i thought key-specs are glued (as key-specs), or that the initial triple can be glued with key-spec. but now i understand that gluing is only needed for the purpose of the backwards compatible output of COMMAND command.
and that the module API for command registration is just for compatibility with old modules, and should be overwritten by key-specs completely (if defined).

however, when key-specs are defined, several specs could be glued into one triple of COMMAND output, that's the same as with redis native commands.
but not glued into the one defined by the command creation triple.

not sure which gaps we still have, but i suppose seeing either the implementation of the tests, we can spot them.

Reply:
i've reviewed the code and tests again (finally bothered to look at the fact and not just talk about them).
first, in contrast to what it was in the past, key-specs are not defined incrementally, they're declared in one go.
when that happens, if there are any range specs, they completely override the legacy spec that's defined in the command creation (and if there's an opportunity to glue several specs into one legacy spec we do that).
if there are no range specs, then whatever was defined at command creation remains in the legacy spec.
the tests cover these cases (the overwrite).

what i think the tests don't cover is:
1. a case where a module doesn't declare key-specs at all. the test can verify that there are no key-specs, but the triple in COMMAND info exists.
2. a case where the module declares only non-range specs, so the legacy range in command creation survived? (i'm not certain if that's the right behavior, we could have also wiped it, but i'm not sure why a module would do such a thing).

p.s. in one of my previous comments i said: "a key-spec that's added in addition to the one in command creation.", i guess that's wrong. we never create key-specs from the initial triple, and when key-specs are defined we should always override that triple if possible, not amend to it.

Reply:
Yes, obviously all key-specs come in one go, in an array. (I though it's so obvious I didn't even explain it, thus confusion.)

I've added the two tests mentioned, without changing any implementation for now. In the second case (only non-range specs), the legacy triple is wiped.  I guess the comment in server.c `` is right actually, and it can be seen from the first line in the function (memset):

```C
/* (...)
 * If this functions fails it means that the legacy (first,last,step)
 * spec used by COMMAND will show 0,0,0. This is not a dire situation
 * because anyway the legacy (first,last,step) spec is to be deprecated
 * and one should use the new key specs scheme.
 */
void populateCommandLegacyRangeSpec(struct redisCommand *c) {
    memset(&c->legacy_range_key_spec, 0, sizeof(c->legacy_range_key_spec));
```

WDYT? Should we avoid clearing the triple for module commands?

Reply:
thanks.
i'm actually ok with that behavior too (as noted in my previous post), maybe i was reading some doc comment that caused me realize it would work otherwise (or maybe it's just that i missed that memset when reading the code).
so from my perspective, it's good that this is defined and tested...